File: src/backend/vendas/src/model/venda_frete_model.py
from src import db
from src.model.view_pessoa_transportadora_model import ViewPessoaTransportadoraModel


class VendaFreteModel(db.Model):
    __tablename__ = 'venda_frete'

    id = db.Column(db.Integer, primary_key=True)
    responsavel = db.Column(db.String(1))
    conhecimento = db.Column(db.Integer)
    placa = db.Column(db.String(7))
    uf_placa = db.Column(db.String(2))
    selo_fiscal = db.Column(db.Integer)
    quantidade_volume = db.Column(db.Float)
    marca_volume = db.Column(db.String(50))
    especie_volume = db.Column(db.String(20))
    peso_bruto = db.Column(db.Float)
    peso_liquido = db.Column(db.Float)
    id_venda_cabecalho = db.Column(db.Integer, db.ForeignKey('venda_cabecalho.id'))
    id_transportadora = db.Column(db.Integer, db.ForeignKey('view_pessoa_transportadora.id'))

    view_pessoa_transportadora_model = db.relationship('ViewPessoaTransportadoraModel', foreign_keys=[id_transportadora])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_venda_cabecalho = data.get('idVendaCabecalho')
        self.id_transportadora = data.get('idTransportadora')
        self.responsavel = data.get('responsavel')
        self.conhecimento = data.get('conhecimento')
        self.placa = data.get('placa')
        self.uf_placa = data.get('ufPlaca')
        self.selo_fiscal = data.get('seloFiscal')
        self.quantidade_volume = data.get('quantidadeVolume')
        self.marca_volume = data.get('marcaVolume')
        self.especie_volume = data.get('especieVolume')
        self.peso_bruto = data.get('pesoBruto')
        self.peso_liquido = data.get('pesoLiquido')

    def serialize(self):
        return {
            'id': self.id,
            'idVendaCabecalho': self.id_venda_cabecalho,
            'idTransportadora': self.id_transportadora,
            'responsavel': self.responsavel,
            'conhecimento': self.conhecimento,
            'placa': self.placa,
            'ufPlaca': self.uf_placa,
            'seloFiscal': self.selo_fiscal,
            'quantidadeVolume': self.quantidade_volume,
            'marcaVolume': self.marca_volume,
            'especieVolume': self.especie_volume,
            'pesoBruto': self.peso_bruto,
            'pesoLiquido': self.peso_liquido,
            'viewPessoaTransportadoraModel': self.view_pessoa_transportadora_model.serialize() if self.view_pessoa_transportadora_model else None,
        }

File: frontend/vendas/lib/app/data/model/venda_frete_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:vendas/app/data/model/model_imports.dart';
import 'package:vendas/app/data/domain/domain_imports.dart';

class VendaFreteModel {
	int? id;
	int? idVendaCabecalho;
	int? idTransportadora;
	String? responsavel;
	int? conhecimento;
	String? placa;
	String? ufPlaca;
	int? seloFiscal;
	double? quantidadeVolume;
	String? marcaVolume;
	String? especieVolume;
	double? pesoBruto;
	double? pesoLiquido;
	ViewPessoaTransportadoraModel? viewPessoaTransportadoraModel;

	VendaFreteModel({
		this.id,
		this.idVendaCabecalho,
		this.idTransportadora,
		this.responsavel,
		this.conhecimento,
		this.placa,
		this.ufPlaca,
		this.seloFiscal,
		this.quantidadeVolume,
		this.marcaVolume,
		this.especieVolume,
		this.pesoBruto,
		this.pesoLiquido,
		this.viewPessoaTransportadoraModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'responsavel',
		'conhecimento',
		'placa',
		'uf_placa',
		'selo_fiscal',
		'quantidade_volume',
		'marca_volume',
		'especie_volume',
		'peso_bruto',
		'peso_liquido',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Responsavel',
		'Conhecimento',
		'Placa',
		'Uf Placa',
		'Selo Fiscal',
		'Quantidade Volume',
		'Marca Volume',
		'Especie Volume',
		'Peso Bruto',
		'Peso Liquido',
	];

	VendaFreteModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idVendaCabecalho = jsonData['idVendaCabecalho'];
		idTransportadora = jsonData['idTransportadora'];
		responsavel = VendaFreteDomain.getResponsavel(jsonData['responsavel']);
		conhecimento = jsonData['conhecimento'];
		placa = jsonData['placa'];
		ufPlaca = VendaFreteDomain.getUfPlaca(jsonData['ufPlaca']);
		seloFiscal = jsonData['seloFiscal'];
		quantidadeVolume = jsonData['quantidadeVolume']?.toDouble();
		marcaVolume = jsonData['marcaVolume'];
		especieVolume = jsonData['especieVolume'];
		pesoBruto = jsonData['pesoBruto']?.toDouble();
		pesoLiquido = jsonData['pesoLiquido']?.toDouble();
		viewPessoaTransportadoraModel = jsonData['viewPessoaTransportadoraModel'] == null ? ViewPessoaTransportadoraModel() : ViewPessoaTransportadoraModel.fromJson(jsonData['viewPessoaTransportadoraModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idVendaCabecalho'] = idVendaCabecalho != 0 ? idVendaCabecalho : null;
		jsonData['idTransportadora'] = idTransportadora != 0 ? idTransportadora : null;
		jsonData['responsavel'] = VendaFreteDomain.setResponsavel(responsavel);
		jsonData['conhecimento'] = conhecimento;
		jsonData['placa'] = placa;
		jsonData['ufPlaca'] = VendaFreteDomain.setUfPlaca(ufPlaca);
		jsonData['seloFiscal'] = seloFiscal;
		jsonData['quantidadeVolume'] = quantidadeVolume;
		jsonData['marcaVolume'] = marcaVolume;
		jsonData['especieVolume'] = especieVolume;
		jsonData['pesoBruto'] = pesoBruto;
		jsonData['pesoLiquido'] = pesoLiquido;
		jsonData['viewPessoaTransportadoraModel'] = viewPessoaTransportadoraModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idVendaCabecalho = plutoRow.cells['idVendaCabecalho']?.value;
		idTransportadora = plutoRow.cells['idTransportadora']?.value;
		responsavel = plutoRow.cells['responsavel']?.value != '' ? plutoRow.cells['responsavel']?.value : '1-Emitente';
		conhecimento = plutoRow.cells['conhecimento']?.value;
		placa = plutoRow.cells['placa']?.value;
		ufPlaca = plutoRow.cells['ufPlaca']?.value != '' ? plutoRow.cells['ufPlaca']?.value : 'AC';
		seloFiscal = plutoRow.cells['seloFiscal']?.value;
		quantidadeVolume = plutoRow.cells['quantidadeVolume']?.value?.toDouble();
		marcaVolume = plutoRow.cells['marcaVolume']?.value;
		especieVolume = plutoRow.cells['especieVolume']?.value;
		pesoBruto = plutoRow.cells['pesoBruto']?.value?.toDouble();
		pesoLiquido = plutoRow.cells['pesoLiquido']?.value?.toDouble();
		viewPessoaTransportadoraModel = ViewPessoaTransportadoraModel();
		viewPessoaTransportadoraModel?.nome = plutoRow.cells['viewPessoaTransportadoraModel']?.value;
	}	

	VendaFreteModel clone() {
		return VendaFreteModel(
			id: id,
			idVendaCabecalho: idVendaCabecalho,
			idTransportadora: idTransportadora,
			responsavel: responsavel,
			conhecimento: conhecimento,
			placa: placa,
			ufPlaca: ufPlaca,
			seloFiscal: seloFiscal,
			quantidadeVolume: quantidadeVolume,
			marcaVolume: marcaVolume,
			especieVolume: especieVolume,
			pesoBruto: pesoBruto,
			pesoLiquido: pesoLiquido,
		);			
	}

	
}